File: frogment/src/main/java/ch/grze/frogment/core/provider/ReflectionFragmentInstanceProvider.java
package ch.grze.frogment.core.provider;

import ch.grze.frogment.frogment.FrogmentInterface;

public class ReflectionFragmentInstanceProvider implements FragmentInstanceProvider {
    @Override
    public FrogmentInterface getInstance(Class<? extends FrogmentInterface> frogmentClass) {
        try {
            return frogmentClass.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            throw new UnableToCreateFrogmentInstanceException(frogmentClass);
        }
    }
}

File: frogment/src/main/java/ch/grze/frogment/core/provider/ReflectionFragmentInstanceProvider.kt
package ch.grze.frogment.core.provider

import ch.grze.frogment.frogment.FrogmentInterface

class ReflectionFragmentInstanceProvider : FragmentInstanceProvider {
    override fun getInstance(frogmentClass: Class<out FrogmentInterface>): FrogmentInterface {
        try {
            return frogmentClass.newInstance()
        } catch (e: Exception) {
            when (e) {
                is InstantiationException, is IllegalAccessException -> throw UnableToCreateFrogmentInstanceException(frogmentClass)
                else -> throw e
            }
        }
    }
}
